Flakiness: timed out waiting for transaction in stress test

https://circleci.com/gh/orbs-network/orbs-network-go/2731

```
orbs-acceptance_1  | 
orbs-acceptance_1  | --- FAIL: TestCreateGazillionTransactionsWhileTransportIsDelayingRandomMessages (1.06s)
orbs-acceptance_1  | panic: timed out waiting for transaction with hash 6b4bcb17c3f0ef4680c6f47cd43dcea63c094425f9c26ca1b37614e2c7793e52 [recovered]
orbs-acceptance_1  | 	panic: timed out waiting for transaction with hash 6b4bcb17c3f0ef4680c6f47cd43dcea63c094425f9c26ca1b37614e2c7793e52
orbs-acceptance_1  | 
orbs-acceptance_1  | goroutine 12505 [running]:
orbs-acceptance_1  | testing.tRunner.func1(0xc42050c690)
orbs-acceptance_1  | 	/usr/local/go/src/testing/testing.go:742 +0x29d
orbs-acceptance_1  | panic(0x98a680, 0xc420152d30)
orbs-acceptance_1  | 	/usr/local/go/src/runtime/panic.go:502 +0x229
orbs-acceptance_1  | github.com/orbs-network/orbs-network-go/test/harness/services/blockstorage/adapter.(*inMemoryBlockPersistence).WaitForTransaction(0xc4204740c0, 0xa8bd00, 0xc420474000, 0xc420694ba0, 0x20, 0x20, 0x3b9aca00, 0xc420695c00)
orbs-acceptance_1  | 	/go/src/github.com/orbs-network/orbs-network-go/test/harness/services/blockstorage/adapter/memory_persistence.go:59 +0x384
orbs-acceptance_1  | github.com/orbs-network/orbs-network-go/test/harness.(*inProcessNetwork).WaitForTransactionInStateForAtMost(0xc420326c30, 0xa8bd00, 0xc420474000, 0x0, 0xc420694ba0, 0x20, 0x20, 0x3b9aca00)
orbs-acceptance_1  | 	/go/src/github.com/orbs-network/orbs-network-go/test/harness/network.go:87 +0xb3
orbs-acceptance_1  | github.com/orbs-network/orbs-network-go/test/harness.(*inProcessNetwork).WaitForTransactionInState(0xc420326c30, 0xa8bd00, 0xc420474000, 0x0, 0xc420694ba0, 0x20, 0x20)
orbs-acceptance_1  | 	/go/src/github.com/orbs-network/orbs-network-go/test/harness/network.go:83 +0x73
orbs-acceptance_1  | github.com/orbs-network/orbs-network-go/test/acceptance.sendTransfersAndAssertTotalBalance(0xa8bd00, 0xc420474000, 0xa8e1c0, 0xc420326c30, 0xc42050c690, 0x64)
```

doesn't look like a deadlock from the logs

--- a/test/harness/services/blockstorage/adapter/memory_persistence.go
+++ b/test/harness/services/blockstorage/adapter/memory_persistence.go
@@ -24,12 +24,18 @@ type InMemoryBlockPersistence interface {
 type blockHeightChan chan primitives.BlockHeight
 
 type inMemoryBlockPersistence struct {
-	blockPairs     []*protocol.BlockPairContainer
+	blockChain struct {
+		sync.RWMutex
+		blocks []*protocol.BlockPairContainer
+	}
+
 	failNextBlocks bool
 	tracker        *synchronization.BlockTracker
 
-	lock                  *sync.RWMutex
-	blockHeightsPerTxHash map[string]blockHeightChan
+	blockHeightsPerTxHash struct {
+		sync.Mutex
+		channels map[string]blockHeightChan
+	}
 }
 
 func NewInMemoryBlockPersistence() InMemoryBlockPersistence {
@@ -37,13 +43,14 @@ func NewInMemoryBlockPersistence() InMemoryBlockPersistence {
 }
 
 func NewInMemoryBlockPersistenceWithBlockTimeout(duration time.Duration) InMemoryBlockPersistence {
-	return &inMemoryBlockPersistence{
+	p := &inMemoryBlockPersistence{
 		failNextBlocks: false,
 		tracker:        synchronization.NewBlockTracker(0, 5, duration),
-
-		lock: &sync.RWMutex{},
-		blockHeightsPerTxHash: make(map[string]blockHeightChan),
 	}
+
+	p.blockHeightsPerTxHash.channels = make(map[string]blockHeightChan)
+
+	return p
 }
 
 func (bp *inMemoryBlockPersistence) GetBlockTracker() *synchronization.BlockTracker {
@@ -51,9 +58,7 @@ func (bp *inMemoryBlockPersistence) GetBlockTracker() *synchronization.BlockTrac
 }
 
 func (bp *inMemoryBlockPersistence) WaitForTransaction(ctx context.Context, txhash primitives.Sha256, atMost time.Duration) primitives.BlockHeight {
-	bp.lock.Lock()
 	ch := bp.getChanFor(txhash)
-	bp.lock.Unlock()
 
 	select {
 	case h := <-ch:
@@ -69,10 +74,8 @@ func (bp *inMemoryBlockPersistence) WriteBlock(blockPair *protocol.BlockPairCont
 		return errors.New("could not write a block")
 	}
 
-	bp.lock.Lock()
-	defer bp.lock.Unlock()
+	bp.addBlockToInMemoryChain(blockPair)
 
-	bp.blockPairs = append(bp.blockPairs, blockPair)
 	bp.tracker.IncrementHeight()
 
 	bp.advertiseAllTransactions(blockPair.TransactionsBlock)
@@ -139,10 +142,13 @@ func (bp *inMemoryBlockPersistence) FailNextBlocks() {
 
 // Is covered by the mutex in WriteBlock
 func (bp *inMemoryBlockPersistence) getChanFor(txhash primitives.Sha256) blockHeightChan {
-	ch, ok := bp.blockHeightsPerTxHash[txhash.KeyForMap()]
+	bp.blockHeightsPerTxHash.Lock()
+	defer bp.blockHeightsPerTxHash.Unlock()
+
+	ch, ok := bp.blockHeightsPerTxHash.channels[txhash.KeyForMap()]
 	if !ok {
 		ch = make(blockHeightChan, 1)
-		bp.blockHeightsPerTxHash[txhash.KeyForMap()] = ch
+		bp.blockHeightsPerTxHash.channels[txhash.KeyForMap()] = ch
 	}
 
 	return ch
@@ -173,8 +179,15 @@ func (bp *inMemoryBlockPersistence) GetLastBlock() (*protocol.BlockPairContainer
 }
 
 func (bp *inMemoryBlockPersistence) getBlockPairs() []*protocol.BlockPairContainer {
-	bp.lock.RLock()
-	defer bp.lock.RUnlock()
+	bp.blockChain.RLock()
+	defer bp.blockChain.RUnlock()
 
-	return bp.blockPairs
+	return bp.blockChain.blocks
+}
+
+func (bp *inMemoryBlockPersistence) addBlockToInMemoryChain(blockPair *protocol.BlockPairContainer) {
+	bp.blockChain.Lock()
+	defer bp.blockChain.Unlock()
+
+	bp.blockChain.blocks = append(bp.blockChain.blocks, blockPair)
 }

